upstairs/src/client.rs

```diff
@@ -1232,341 +1232,149 @@ impl DownstairsClient {
-                        ackable = true;
-                        cdt::up__to__ds__barrier__done!(|| ds_id.0);
+                    IOop::Flush { .. } => {
+                        assert!(read_data.blocks.is_empty());
+                        assert!(read_data.data.is_empty());
+                        assert!(extent_info.is_none());
+
+                        self.last_flush = ds_id;
                     }
-                }
-                IOop::ExtentFlushClose { .. } => {
-                    assert!(read_data.blocks.is_empty());
-                    assert!(read_data.data.is_empty());
+                    IOop::ExtentFlushClose { .. } => {
+                        assert!(read_data.blocks.is_empty());
+                        assert!(read_data.data.is_empty());
 
-                    let ci = self.repair_info.replace(extent_info.unwrap());
-                    if ci.is_some() {
-                        panic!(
+                        let ci = self.repair_info.replace(extent_info.unwrap());
+                        if ci.is_some() {
+                            panic!(
                             "[{}] Unexpected repair found on insertion: {:?}",
                             self.client_id, ci
                         );
+                        }
                     }
-
-                    if jobs_completed_ok == 3 {
-                        debug!(self.log, "ExtentFlushClose {ds_id} AckReady");
-                        ackable = true;
+                    IOop::ExtentLiveRepair { .. }
+                    | IOop::ExtentLiveReopen { .. }
+                    | IOop::ExtentLiveNoOp { .. } => {
+                        assert!(read_data.blocks.is_empty());
+                        assert!(read_data.data.is_empty());
+                        assert!(extent_info.is_none());
                     }
                 }
-                IOop::ExtentLiveRepair { .. } => {
-                    assert!(read_data.blocks.is_empty());
-                    assert!(read_data.data.is_empty());
-                    if jobs_completed_ok == 3 {
-                        debug!(self.log, "ExtentLiveRepair AckReady {ds_id}");
-                        ackable = true;
+                IOState::Done
+            }
+            Err(e) => {
+                // The downstairs sent us this error
+                error!(
+                    self.log,
+                    "DS Reports error {e:?} on job {}, {:?} EC", ds_id, job,
+                );
+                match (&job.work, &e) {
+                    // Some errors can be returned without considering the
+                    // Downstairs bad. For example, it's still an error if a
+                    // snapshot exists already but we should not increment
+                    // `downstairs_errors` and transition that Downstairs to
+                    // Failed - that downstairs is still able to serve IO.
+                    (_, CrucibleError::SnapshotExistsAlready(..)) => (),
```

Comment:
Not matching on Flush here?

Reply:
This matches the previous behavior ([L1296](https://github.com/oxidecomputer/crucible/pull/1561/files#diff-df991e179692a549eace0bb472a6b3c07483ab18527ee9b30f433142a95fd89fL1296)), but I agree that we could be more specific; fixed!

(Note that this only handles error counters, the actual [error checking in the Downstairs](https://github.com/oxidecomputer/crucible/blob/f3d07aa414081b77e13a2f5f0da7bfbaa1749372/upstairs/src/downstairs.rs#L3361-L3363) remains agnostic to `IOop` type)